pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,1100 @@
+			return nil, nil, err
+		}
+	}
+
+	sc, err := scaledContext(restConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	adConfig, err := adConfiguration(sc)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return sc, adConfig, nil
+}
+
+// UnmigrateAdGUIDUsersOnce will ensure that the migration script will run only once.  cycle through all users, ctrb, ptrb, tokens and migrate them to an
+// appropriate DN-based PrincipalID.
+func UnmigrateAdGUIDUsersOnce(sc *config.ScaledContext) error {
+	migrationConfigMap, _ := sc.Core.ConfigMaps(activedirectory.StatusConfigMapNamespace).GetNamespaced(activedirectory.StatusConfigMapNamespace, activedirectory.StatusConfigMapName, metav1.GetOptions{})
+	if migrationConfigMap != nil {
+		migrationStatus := migrationConfigMap.Data[activedirectory.StatusMigrationField]
+		switch migrationStatus {
+		case activedirectory.StatusMigrationFinished:
+			logrus.Debugf("[%v] ad-guid migration has already been completed, refusing to run again at startup", migrateAdUserOperation)
+			return nil
+		}
+	}
+	return UnmigrateAdGUIDUsers(&sc.RESTConfig, false, false)
+}
+
+// UnmigrateAdGUIDUsers will cycle through all users, ctrb, ptrb, tokens and migrate them to an
+// appropriate DN-based PrincipalID.
+func UnmigrateAdGUIDUsers(clientConfig *restclient.Config, dryRun bool, deleteMissingUsers bool) error {
+	if dryRun {
+		logrus.Infof("[%v] dryRun is true, no objects will be deleted/modified", migrateAdUserOperation)
+		deleteMissingUsers = false
+	} else if deleteMissingUsers {
+		logrus.Infof("[%v] deleteMissingUsers is true, GUID-based users not present in Active Directory will be deleted", migrateAdUserOperation)
+	}
+
+	sc, adConfig, err := prepareClientContexts(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	migrationConfigMap, _ := sc.Core.ConfigMaps(activedirectory.StatusConfigMapNamespace).GetNamespaced(activedirectory.StatusConfigMapNamespace, activedirectory.StatusConfigMapName, metav1.GetOptions{})
+	if migrationConfigMap != nil {
+		migrationStatus := migrationConfigMap.Data[activedirectory.StatusMigrationField]
+		switch migrationStatus {
+		case activedirectory.StatusMigrationRunning:
+			logrus.Infof("[%v] ad-guid migration is currently running, refusing to run again concurrently", migrateAdUserOperation)
+			return nil
+		}
+	}
+
+	// We'll share this lConn for all lookups to hopefully speed things along
+	lConn, err := ldapConnection(adConfig)
+	if err != nil {
+		return err
+	}
+	defer lConn.Close()
+
+	err = updateMigrationStatus(sc, activedirectory.StatusMigrationField, activedirectory.StatusMigrationRunning)
+	if err != nil {
+		return fmt.Errorf("unable to update migration status configmap: %v", err)
+	}
+
+	users, err := sc.Management.Users("").List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to fetch user list: %v", err)
+	}
+
+	usersToMigrate, missingUsers, skippedUsers := identifyMigrationWorkUnits(users, lConn, adConfig)
```

Comment:
If there are no users to migrate (the first return option) should the script write the status is finished and exit early?

Reply:
Not here, because there might be missingUsers and skippedUsers that need to be logged. In particular, a misconfigured Active Directory provider might cause **all** users to end up in either the missing (couldn't find any GUIDs) or skipped (couldn't connect to LDAP at all) buckets. We want to report these errors, and the following logic does just that.